framework/auth/views.py

```diff
@@ -202,87 +202,157 @@ def forgot_password_post(auth, **kwargs):
-        next_url = campaigns.campaign_url_for(campaign)
 
-    if not next_url:
-        next_url = request.args.get('redirect_url')
-        must_login_warning = False
+def validate_next_url(next_url):
+    """
+    Non-view helper function that checks `next_url`.
+    Only allow redirects which are relative root or full domain (CAS, OSF and MFR).
+    Disallows external redirects.
 
-    if not validate_next_url(next_url):
-        raise HTTPError(http.BAD_REQUEST)
+    :param next_url: the next url to check
+    :return: True if valid, False otherwise
+    """
 
-    if auth.logged_in:
-        if not log_out:
-            if next_url:
-                return redirect(next_url)
-            return redirect('dashboard')
-        # redirect user to CAS for logout, return here w/o authentication
-        return auth_logout(redirect_url=request.url)
+    # disable external domain using `//`: the browser allows `//` as a shortcut for non-protocol specific requests
+    # like http:// or https:// depending on the use of SSL on the page already.
+    if next_url.startswith('//'):
+        return False
+    # only OSF, MFR and CAS domains are allowed
+    if not (next_url[0] == '/' or
+            next_url.startswith(settings.DOMAIN) or
+            next_url.startswith(settings.CAS_SERVER_URL) or
+            next_url.startswith(settings.MFR_SERVER_URL)):
+        return False
+    return True
+
+
+@collect_auth
+def auth_register(auth):
+    """
+    View for OSF register. Land on the register page, redirect or go to `auth_logout`
+    depending on `data` returned by `login_and_register_handler`.
 
-    status_message = request.args.get('status', '')
-    if status_message == 'expired':
-        status.push_status_message('The private link you used is expired.', trust=False)
-        status.push_status_message('The private link you used is expired.  Please <a href="/settings/account/">'
-                                   'resend email.</a>', trust=False)
+    :param auth: the auth context
+    :return: land, redirect or `auth_logout`
+    :raise: http.BAD_REQUEST
+    """
+
+    context = {}
+    # a target campaign in `auth.campaigns`
+    campaign = request.args.get('campaign')
+    # the service url for CAS login or redirect url for OSF
+    next_url = request.args.get('next')
+    # used only for `claim_user_registered`
+    logout = request.args.get('logout')
 
-    if next_url and must_login_warning:
-        status.push_status_message(language.MUST_LOGIN, trust=False)
+    data = login_and_register_handler(auth, login=False, campaign=campaign, next_url=next_url, logout=logout)
 
-    # set login_url to form action, upon successful authentication specifically w/o logout=True,
-    # allows for next to be followed or a redirect to the dashboard.
-    redirect_url = web_url_for('auth_login', next=next_url, _absolute=True)
+    # land on register page
+    if data['status_code'] == http.OK:
+        if data['must_login_warning']:
```

Comment:
this `must_login_warning` shows up only if it is the register page and if the flag is on, there is a `return` for each top level conditional